atuin-common/src/utils.rs

```diff
@@ -100,6 +102,34 @@ pub fn is_bash() -> bool {
     env::var("ATUIN_SHELL_BASH").is_ok()
 }
 
+/// Extension trait for anything that can behave like a string to make it easy to escape control
+/// characters.
+///
+/// Intended to help prevent control characters being printed and interpreted by the terminal when
+/// printing history as well as to ensure the commands that appear in the interactive search
+/// reflect the actual command run rather than just the printable characters.
+pub trait Escapable: AsRef<str> {
+    fn escape_control(&self) -> Cow<str> {
+        if !self.as_ref().contains(|c: char| c.is_ascii_control()) {
+            self.as_ref().into()
+        } else {
+            let mut part = self.as_ref();
+            // Not a perfect way to reserve space but should reduce the allocations
+            let mut buf = String::with_capacity(part.as_bytes().len());
+            while let Some(i) = part.find(|c: char| c.is_ascii_control()) {
+                // safe to index with `..i`, `i` and `i+1..` as part[i] is a single byte ascii char
+                buf.push_str(&part[..i]);
+                let _ = write!(&mut buf, "\\x{:02x}", part.as_bytes()[i]);
```

Comment:
This might be just a preference, but is there a possibility to use the caret notation instead of the hex notation `\xHH`? I'm more familiar with the caret notation in the command line and thus the command strings (as the shells' line editors show the control characters in the caret notation). (Note: I'm not a maintainer, so I don't have any power. If you don't like this, you can ignore it.)

```suggestion
                buf.push('^');
                buf.push(match part.as_bytes()[i] {
                    0x7F => '?',
                    code => char::from_u32(u32::from(code) + 64).unwrap(),
                });
```

Reply:
+1 here

Reply:
> but the main problem with using `^[` etc is that you can't copy/paste the output to re-run the command as they'll be interpreted as two characters instead of one.

Ah, OK. I now understand the third paragraph of the cover of this PR. Maybe, is `echo '[31mtwo'` in the cover a typo of `echo '\x1b[31mtwo'`?

But then, I'm afraid to say that the hex notation still has the same problem for copying and pasting as the caret notation. The raised example of `echo` works with the hex notation because of the specific behavior of a command of a specific shell (i.e., echo in Bash and Fish doesn't work that way). The hex notation is not handled at the shell level, so we cannot expect it to work in general. I think there are even opposite cases, where the caret notation works while the hex notation fails. For example, in Zsh, `bindkey ^R some_widget` works, but `bind \x0D some_widget` doesn't work as expected.

In the first place, the users can also copy and paste the commands in the terminal's scrollback buffer, where the control characters in the commands are presented with the caret notation. I'm not sure if using the hex notation in Atuin would be actually useful in solving the essential issue. If a user wants to include a raw control character in commands, I feel the user finally needs to learn that copying and pasting the displayed command doesn't work in general.

Reply:
I think I'm coming round to this idea more now. It's less convenient for the cases I had that originally come across but I can see the `^[` version is more general and better matches the way the terminal displays control chars. As Ellie initially `+1`ed your original comment I'll change it now. Is it worth inverting the colours for escape characters (similar to how zsh does it) to distinguish it from the literal `^` + `[`?

Reply:
> Is it worth inverting the colours for escape characters (similar to how zsh does it) to distinguish it from the literal `^` + `[`?

Yeah, that sounds better. However, we might need adjustments for the ongoing PR #1574, which uses the terminal's reverse video for another meaning. In principle, we can treat *(reverse + reverse)* as *not-reverse*, but the terminal control sequences do not offer such a model. Someone needs to implement it in the application layer, but I'm not sure if that is already implemented in any of the Rust libraries that Atuin currently depends on.

Reply:
Ok, I'll leave the reverse for now, it can be figured out whenever 1574 is merged.

Reply:
> > It could be possible to use caret notation for the interactive search but the hex version when things are printed to stdout but I'm not sure the inconsistency would be worth it.
> 
> I think we usually switch the output format depending on whether `stdout` (or the destination stream) is connected to a TTY. We can escape the control characters when `stdout` is connected to TTY, or otherwise, output the raw command string.

Do you have a plan to fix this issue? The current `main` branch doesn't work for the multiline command:

![image](https://github.com/atuinsh/atuin/assets/8982192/e847e10f-2111-4512-8780-0e253d4f4658)

`\x0a` is not treated as a newline of the shell syntax.

Reply:
Then yes, looks like I need to. Would it be better to add the tty check or should the search result always be printed as is? Or a flag could switch between the two (although I'd accept if there's a reluctance to add more flags to search)?

Reply:
Hmm, I leave it to you. It's similar to the coloring of the output of the commands. Some commands do not support colors at all. Some commands enable colors with the check for TTY through `--color=auto`. Some commands perform the TTY switching by default. All are understandable.

Reply:
#1585 and #1586 for these two changes.

Reply:
are they both ok for review? I see they're marked as drafts

Reply:
Sure, I just figured I should test these changes more thoroughly than I did for this one and didn't have time when I submitted them. Given main is currently broken though I thought it was worth getting something in.

Reply:
Thanks, I'll have a look at what changed. And I will do but it might not be for a while.